internal/uvm/vpmem_mapped.go

```diff
@@ -0,0 +1,308 @@
+					"deviceOffset": vhd.mappedRegion.Offset(),
+				}).Debug("found mapped VHD")
+				return i, vhd, nil
+			}
+		}
+	}
+	return 0, nil, ErrNotAttached
+}
+
+// allocateNextVPMemMappedDeviceLocation allocates a memory region with a minimum offset on the VPMem surface,
+// where the device with a given `devSize` can be mapped.
+func (uvm *UtilityVM) allocateNextVPMemMappedDeviceLocation(ctx context.Context, devSize uint64) (uint32, memory.MappedRegion, error) {
+	// device size has to be page aligned
+	devSize = pageAlign(devSize)
+
+	for i := uint32(0); i < uvm.vpmemMaxCount; i++ {
+		pmem := uvm.vpmemDevicesMultiMapped[i]
+		if pmem == nil {
+			pmem = newPackedVPMemDevice()
+			uvm.vpmemDevicesMultiMapped[i] = pmem
+		}
+
+		if len(pmem.mappings) >= int(pmem.maxMappedDeviceCount) {
+			continue
+		}
+
+		reg, err := pmem.Allocate(devSize)
+		if err != nil {
+			continue
+		}
+		log.G(ctx).WithFields(logrus.Fields{
+			"deviceNumber": i,
+			"deviceOffset": reg.Offset(),
+			"deviceSize":   devSize,
+		}).Debug("found offset for mapped VHD on an existing VPMem device")
+		return i, reg, nil
+	}
+	return 0, nil, ErrNoAvailableLocation
+}
+
+// addVPMemMappedDevice adds container layer as a mapped device, first mapped device is added as a regular
+// VPMem device, but subsequent additions will call into mapping APIs
+//
+// Lock MUST be held when calling this function
+func (uvm *UtilityVM) addVPMemMappedDevice(ctx context.Context, hostPath string) (_ string, err error) {
+	if _, dev, err := uvm.findVPMemMappedDevice(ctx, hostPath); err == nil {
+		dev.refCount++
+		return dev.uvmPath, nil
+	}
+
+	devSize, err := fileSystemSize(hostPath)
+	if err != nil {
+		return "", err
+	}
+	deviceNumber, memReg, err := uvm.allocateNextVPMemMappedDeviceLocation(ctx, devSize)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			pmem := uvm.vpmemDevicesMultiMapped[deviceNumber]
+			if err := pmem.Release(memReg); err != nil {
+				log.G(ctx).WithError(err).Debugf("failed to reclaim pmem region: %s", err)
+			}
+		}
+	}()
+
+	uvmPath := fmt.Sprintf(lcowPackedVPMemLayerFmt, deviceNumber, memReg.Offset(), devSize)
+	md := newVPMemMappedDevice(hostPath, uvmPath, devSize, memReg)
+	modification, err := newMappedVPMemModifyRequest(ctx, requesttype.Add, deviceNumber, md, uvm)
+	if err := uvm.modify(ctx, modification); err != nil {
+		return "", errors.Errorf("uvm::addVPMemMappedDevice: failed to modify utility VM configuration: %s", err)
+	}
+	defer func() {
+		if err != nil {
+			rmRequest, _ := newMappedVPMemModifyRequest(ctx, requesttype.Remove, deviceNumber, md, uvm)
+			if err := uvm.modify(ctx, rmRequest); err != nil {
+				log.G(ctx).WithError(err).Debugf("failed to rollback modification")
+			}
+		}
+	}()
+
+	pmem := uvm.vpmemDevicesMultiMapped[deviceNumber]
+	if err := pmem.mapVHDLayer(ctx, md); err != nil {
+		return "", errors.Wrapf(err, "failed to update internal state")
+	}
+	uvm.vpmemDevicesMultiMapped[deviceNumber] = pmem
```

Comment:
pmem is a pointer though so wouldn't the entry in `vpmemDevicesMultiMapped` get updated already?